Signup: Create new social account when signing up with Google

drewblaisdell opened the pull request:
Requires D4835-code.

This PR:

- Adds `propTypes` to `SignupForm`.
- Updates signup to create a new social account when the user signs up with Google.
- Moves the global notice from the user step closer to the social buttons. This is rarely ever displayed since we validate normal signups, but is also displayed in the case that the user attempts to sign up with a Google account that has the same email address as an existing non-social account.
- Removes the error that is thrown when the last step in signup has errors. This is not really an error with our application, so simply handling the redirect properly (i.e. not redirecting, because that error occurs on the last step) is adequate. It also removes the code to reset the stores if the last step contains an error, since the user might want to fix an issue with existing data.

**Testing**
*New social signup*
- Create a [new Google account](https://accounts.google.com/signup?hl=en).
- Visit [`/start/social`](http://calypso.localhost:3000/start/social).
- Click `Google` and authorize the application with the account you created.
- Assert that you are signed up properly and are logged in after clicking `Continue`.

*Attempting to sign up with an existing social account*
- Visit `/start/social`.
- Click `Google` and authorize the application with the account you created in the previous step.
- Assert that you are signed *in* properly after clicking `Continue`. We may want different copy here to let the user know that we're logging them into their existing account.

*Attempting to sign up with an existing non-social account*
- Visit `/start/social`.
- Click `Google` and authorize a Google account that has an email address for a non-social WordPress.com account. For me, this was my main Google account, as I have an existing WordPress.com account for `drew.blaisdell@gmail.com`.
- Assert that you are redirected to the user step with an error indicating that this account already exists. We may want to change the copy here or provide a link for the user to easily log in to this account.

- [x] Code
- [x] Product

drewblaisdell committed:
Components: Add `propTypes` to `SignupForm`

drewblaisdell committed:
Signup: Create social account when the user signs up with Google

drewblaisdell committed:
Signup: Move global error notice closer to social buttons in `SignupForm`

drewblaisdell committed:
Signup: Remove error/store reset when there is an error on the current step

In this situation, the user might simply need to fix a value in an existing input. We can simply not redirect the user to the same step if they are already on an invalid step.

drewblaisdell commented:
I have a couple questions for @Automattic/cerulean and @fabianapsimoes:

- What should the behavior be when the user attempts to sign up with an existing social account? Currently, the patch/PR logs the user in to their existing social WordPress.com account if they attempt to do this, which will make more sense in the context of creating a new site through `/start` (currently `/start/social` just creates an account), but we might want to message to the user that they are signing into an existing account they already created.
- What should we display when the user tries to sign up with a social account associated with an existing, non-social WordPress.com account (i.e. one that cannot be signed in to with Google auth)? Currently, an error message is displayed:

<img width="415" alt="screen shot 2017-03-21 at 3 43 45 pm" src="https://cloud.githubusercontent.com/assets/1130674/24174693/35f33a42-0e4f-11e7-9a14-be9cb9a75b72.png">

but at the very least, I think we want a saner error message, and to link the user to the log in form for their email address. We may also want to allow the user to log in and connect their social account with their existing WordPress.com account (@scruffian mentioned this being as something we may want), in which case we should probably handle it separately.

ranh commented:
> What should the behavior be when the user attempts to sign up with an existing social account?

What do you mean by "existing social account"?

If users choose Google on the sign up screen, and their Google account is already associated with an account on WordPress.com (for example, if they used it to sign up originally), then they should be logged in. 

I agree that showing a notice to confirm the result (for e.g., "Logging in to your account") could make sense. But this probably depends on how this is integrated into an overall flow. For example, if you're signing up (or logging in) in order to like a post, there's probably no need to go into any detail. 

> What should we display when the user tries to sign up with a social account associated with an existing, non-social WordPress.com account (i.e. one that cannot be signed in to with Google auth)?

To clarify, this is when users authorize their social account, but the email address on that account is already in use on wpcom?

If so, then as you mentioned, we would want to handle this by letting users connect with the existing account after returning to wpcom. Not sure where this fits in the roadmap, for now perhaps it's enough to just tweak the notice a bit? 

> It looks like you already have an account on WordPress.com. [Log in](#).

Does it make sense to specify the email? Since users are already authenticated it sounds like it should be fine, and it would make the issue much clearer. We can say "Log in with mail@example.com" instead of just "Log in". And even pre-populate the email field on the log in screen?

josemarques commented:
On the [prototype](https://invis.io/WHARLW9RU) where we tested the sign up flow we started from an email / social networks form that acted as both sign up and log in:

![screen shot 2017-03-22 at 10 12 02](https://cloud.githubusercontent.com/assets/57050/24192778/06fd3b22-0ee8-11e7-8c4b-2d7f1006a9be.png)

If the user started with an email of an existing WordPress.com account or tried to sign up with a social profile using an email of an account that already existed we display this:

![screen shot 2017-03-22 at 10 14 19](https://cloud.githubusercontent.com/assets/57050/24192887/5af2b6c6-0ee8-11e7-962b-344d9b40b3d9.png)

The message does not follow the Calypso style for notices since it is instructional and not really a notice.

fabianapsimoes commented:
@drewblaisdell:

>  We may also want to allow the user to log in and connect their social account with their existing WordPress.com account (@scruffian mentioned this being as something we may want), in which case we should probably handle it separately.

Agreed, but this will probably need some design work from Cerulean (cc @breezyskies). We need to decide where in `/me` we want this option to live. As Ran mentions, this is probably work for a future milestone, and for now we should just tweak the copy for the notice.

Also, product looks good once the necessary copy tweaks are made. 

Really great work with this PR / patch, Drew :cake: 

@ranh: 

> What do you mean by "existing social account"?

I believe a social account is a WordPress.com account created via Google / Facebook authentication. Accounts created using the form (email / username / password) would be non social accounts.

> I agree that showing a notice to confirm the result (for e.g., "Logging in to your account") could make sense. But this probably depends on how this is integrated into an overall flow. For example, if you're signing up (or logging in) in order to like a post, there's probably no need to go into any detail.

Right, I agree. But how should we handle this for the signup flow, which is what we're focusing on with this PR (part of M1)?

For reference, this is how we handle the case where a user enters an existing email address for a non-social account: 

![screenshot-calypso localhost-3000-2017-03-22-11-13-30 1](https://cloud.githubusercontent.com/assets/1248436/24193048/312c262a-0ef1-11e7-9262-7b57e11fd384.png)

We should probably do something similar here?

hafizrahman commented:
All three cases are working fine.

I find that for the second case, sometimes Google does not ask to authenticate (probably because it knows I'm already logged in to Gmail in another tab? Or because something is still cached somewhere?). Instead what happens is that the auth pop-up shows up real quick then disappears, then I am transferred to the screen with the "Continue" button, after which I can log in properly.

So it's not an issue, really, just something to note.

ranh commented:
> Right, I agree. But how should we handle this for the signup flow, which is what we're focusing on with this PR (part of M1)?

For `/start/`, I don't think anything more is needed. Users should just proceed with creating their account, whether they authenticated for an existing account or we created a new one for them. 

I think this will be more about what we say on the account creation screen itself (before users choose Google or any other option). For example, probably no need to refer to "sign up" or "log in" specifically.

As suggested above, we can say something like "Logging in to your account" (or "Creating your account"), especially if we have some interstitial or loading screen anyway. 

> For reference, this is how we handle the case where a user enters an existing email address for a non-social account

Was this re. the second question? We can be more straightforward in this case, since we can assume that the owner of the Google account also owns the existing wpcom account.

Maybe we just take them to log in directly?

ranh commented:
> For /start/, I don't think anything more is needed

But would love to hear what the rest of @Automattic/cerulean thinks

kristastevens commented:
Drew B. asked: 

> What should we display when the user tries to sign up with a social account associated with an existing, non-social WordPress.com account (i.e. one that cannot be signed in to with Google auth)? Currently, an error message is displayed: 

> but at the very least, I think we want a saner error message, and to link the user to the log in form for their email address. 

I think @ranh makes a great suggestion here and this is what I would advocate for. The language is clear, concise, and jargon-free: 

> It looks like you already have an account on WordPress.com. Log in.  We can say "Log in with mail@example.com" instead of just "Log in". And even pre-populate the email field on the log in screen

I think pre-populating the email field is an important step here and a nice service to the user.

scruffian commented:
> What should the behavior be when the user attempts to sign up with an existing social account? Currently, the patch/PR logs the user in to their existing social WordPress.com account if they attempt to do this, which will make more sense in the context of creating a new site through /start (currently /start/social just creates an account), but we might want to message to the user that they are signing into an existing account they already created.

I think this is fine, a message might be useful for clarification.

> What should we display when the user tries to sign up with a social account associated with an existing, non-social WordPress.com account (i.e. one that cannot be signed in to with Google auth)? Currently, an error message is displayed:

IMO we shouldn't treat this any differently to the case above, we should log them in to their existing WPCOM account. I don't think we should require users to log in before making this connection.

danhauk commented:
> For /start/, I don't think anything more is needed. Users should just proceed with creating their site, whether they authenticated for an existing account or we created a new one for them.

@ranh I agree here. No need to complicate things, since the idea is to make sign up and log in smoother.

> I don't think we should require users to log in before making this connection.

@scruffian Are there security concerns with this approach? For example, if I log out of my WordPress.com account because I don't want anyone messing with my super important site, but I'm still logged into my Google or Facebook account. Someone could log in to my WordPress.com account without knowing my password.

ranh commented:
> IMO we shouldn't treat this any differently to the case above, we should log them in to their existing WPCOM account. I don't think we should require users to log in before making this connection.

This sounds a bit risky. When users choose the social option on the sign up screen, they don't necessarily have their existing wpcom account in mind (in fact they probably don't have it in mind, or they would use it to log in). So they may not fully understand what they are agreeing to. Adding a step to indicate that they are connecting their Facebook account with their existing wpcom account would make this clearer. And possibly allow users to abort the social sign up and simply log in directly.

, since I'm not sure it will be clear enough to users that by choosing a social option on the sign up screen, they are agreeing to connect it with an existing account that they may not even be aware of at

scruffian commented:
> Adding a step to indicate that they are connecting their social account with their existing wpcom account would make this clearer. And possibly allow users to abort the social sign up and simply log in directly.

This is a good idea. My point was more that I don't think we need to require users to log in before their account is connected.

> Are there security concerns with this approach? For example, if I log out of my WordPress.com account because I don't want anyone messing with my super important site, but I'm still logged into my Google or Facebook account. Someone could log in to my WordPress.com account without knowing my password.

This is true, but its also the case today if a user leaves themselves logged in to their email account on the same computer - I can access you account if I have access to your email address.

stephanethomas commented on client/lib/signup/step-actions.js:
```diff
@@ -236,31 +236,44 @@ module.exports = {
+				if ( errors ) {
+					callback( errors );
+				} else {
+					callback( null, response );
```
Super minor but I see we pass `undefined` instead of `null` to all other callbacks in this component. Maybe we should keep doing it that way here to stay consistent?

stephanethomas commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -1446,6 +1446,29 @@ Undocumented.prototype.usersNew = function( query, fn ) {
+ *
+ * @return {Promise} A promise for the request
+ */
+Undocumented.prototype.usersSocial = function( service, token, fn ) {
```
I think this function should be named `usersSocialNew`.

stephanethomas commented:
I had two very minor comments but otherwise the code looks great.

I've tested flows from the testing instructions as well the regular signup flow and everything worked fine.

I think it would be a good idea to run E2E tests before merging this pull request.

drewblaisdell commented:
Thanks for all of the reviews, everyone.

I updated our internal list of development tasks for M2 to include updating the notice for existing accounts to link to `Log in` per @ranh's comment, because if we implemented that now, we'd need to change it once the new page exists.

I have also deployed the patch and asserted that this passes the e2e tests per @stephanethomas's comment.
